### ISSUE TITLE
How to implement a Yield

### ISSUE THREAD

**Issue by taooceros** [monoio-rs/monoio#322]
**Version**
0.2.4

**Platform**
Linux seoul-microk8s 5.15.0-125-generic

**Description**
I need to do a customized polling

```rust
pub async fn polling(self: Rc<Self>) {
    unsafe {
        while Rc::strong_count(&self) > 1 {
            while doca_pe_progress(self.pe.as_ptr()) > 0 {
                println!("progress");
            }
            futures_lite::future::yield_now().await;
        }
    }
}
```

How could I implement a yield in monoio? If I follow the standard way (e.g. like futures_lite)

```rust
pub fn yield_now() -> YieldNow {
    YieldNow(false)
}

/// Future for the [`yield_now()`] function.
#[derive(Debug)]
#[must_use = "futures do nothing unless you `.await` or poll them"]
pub struct YieldNow(bool);

impl Future for YieldNow {
    type Output = ();

    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
        if !self.0 {
            self.0 = true;
            cx.waker().wake_by_ref();
            Poll::Pending
        } else {
            Poll::Ready(())
        }
    }
}
```

Because it notify the waker, it will be inserted into the front of the task queue instead of the back of the task queue, which means other tasks will be starved (and then because nobody submit a task it will just poll forever instead of doing anything valuable).

https://github.com/bytedance/monoio/blob/704e7d594b3a7c95f6ae692c0b8cdfee3aebf246/monoio/src/scheduler.rs#L12-L14

The same version works in `tokio`, which can either due to their scheduling policy or the cooperative yielding.

In `tokio` special version of `yield_now`, they includes a `context::defer(cx.waker());`, which probably prevent the needs for using the cooperative yielding.

**Comment by harry-van-haaren** [comment]
### +1 to this request to add `yield_now()` 
The `yield_now()` function is very useful, and missing from Monoio today.

I've dug into the Tokio sources a bit, and there is a seperate way of tracking `&Waker` instances and scheduling those, without having the corresponding `Future` instance. This is how tokio schedules the Waker, but doesn't `spawn()` or do any handling of the `Future` itself. I did not find similar concepts in the MonoIO runtime etc, so I'm not sure the `defer(waker: &Waker)` function can be easily built in MonoIO: perhaps some of the developers closer to the code have a better/novel idea to schedule a `&Waker` in MonoIO.

### A bad but easy solution
A terrible workaround is to do a sleep; it gives very bad performance, but **works** at least. Note that the sleep scheduling is at 1 ms granularity, so changing that to `nanos(1)` isn't going to go any faster :/
```rust
monoio::time::sleep(std::time::Duration::from_millis(1)).await;
```

### The ping-pong channel solution
A better workaround is to use a what I'll call a "ping pong" task, which uses a secondary task to "unpause" the main task again. Some pseudo code might help show how this can be implemented. Before the `loop { /* long running, no .await */ }` you create a `local_sync` (monoio depends on it too) bounded channel, which has async send/recv as its bounded.
```rust
let (yield_tx, mut yield_rx) = local_sync::mpsc::bounded::channel::<()>(1);
```
Spawn a task which only transmits events forever (fixup code & handle errors in this part):
```rust
monoio::spawn(async {
loop {
   yield_tx.send(()).await;
}
});
```
Now in the main task's "forever" loop without `.await` points (where you would otherwise put the `yield_now()` call using Tokio)
```rust
yield_rx.recv().await;
```
Some nice properties of this solution:
- There is no allocation on the "hot path" (the `spawn()` function does a lot of work in the background, so avoid it!)
- The channel is bounded to 1, so the tx task sleeps until the main task has run. At that point, the TX side becomes "writeable" again, which will "unpark" or resume the `yield_tx` side, and cause it to write one more `()` into the channel.
- As we're using a local-channel, there's no atomics/heavyweight instructions here.

A quick check on performance with `perf record` does not show a hotspot on the "ping pong" task code, but I have not done detailed investigation.

### Summary
The ping-pong task is probably OK as a "normal" workaround, however requires implementing at each `yield_now()` location. Perhaps this is acceptable, but I hope `yield_now()` functionality is included as a `pub` function in MonoIO's future.

**Comment by RTEnzyme** [comment]
I implement it by:

```rust
use futures::future::poll_fn;

pub async fn yield_now() {
    monoio::spawn(async move {
        poll_fn(|cx| {
            cx.waker().wake_by_ref();
            Poll::Ready(())
        }).await;
    }).await;
}
```

**Comment by harry-van-haaren** [comment]
Yes that will work, thanks for sharing your solution. Note that it will cause an allocation of the `Future`, and other `spawn()` overhead. If performance matters, the PingPong channel solution is better.

**Comment by RTEnzyme** [comment]
> Yes that will work, thanks for sharing your solution. Note that it will cause an allocation of the `Future`, and other `spawn()` overhead. If performance matters, the PingPong channel solution is better.

Thanks for your reply! But the PingPong channel solution also need to allocate of the `Future`? 
```rust
monoio::spawn(async {
loop {
   yield_tx.send(()).await;
}
});
```

**Comment by harry-van-haaren** [comment]
Correct, but count the number of times it will allocate: in the PingPong solution, the `loop {}` is an infinite loop, so `spawn()` is called once. This results in one allocation of a Future, and then 1000s of `yield_tx().send(())` can occur. With one Future allocated (and in a loop waiting for the other side to wake it..) we have only the cost of the `Waker::wake()` for each `yield_now()`.

In the yield_now() function, *every call* to yield_now() results in a Future being allocated, the poll_fn capturing the context, creating a Waker and wake by ref (inserting the Task back into the run queue of the reactor), and then returning Ready. Note that the allocation of a single Future will be cheap in an unloaded system, particularly when benchmarked in a hot benchmarking loop of  `yield_now()` calls. In production environments, the difference between 1x allocation vs "per yield_now() call allocation" is huge.

**Comment by RTEnzyme** [comment]
> Correct, but count the number of times it will allocate: in the PingPong solution, the `loop {}` is an infinite loop, so `spawn()` is called once. This results in one allocation of a Future, and then 1000s of `yield_tx().send(())` can occur. With one Future allocated (and in a loop waiting for the other side to wake it..) we have only the cost of the `Waker::wake()` for each `yield_now()`.
> 
> In the yield_now() function, _every call_ to yield_now() results in a Future being allocated, the poll_fn capturing the context, creating a Waker and wake by ref (inserting the Task back into the run queue of the reactor), and then returning Ready. Note that the allocation of a single Future will be cheap in an unloaded system, particularly when benchmarked in a hot benchmarking loop of `yield_now()` calls. In production environments, the difference between 1x allocation vs "per yield_now() call allocation" is huge.

LGTM! However, in my workload, the situation where `yield_now` is called is relatively rare; thus, compared to allocating the PingPong channel in advance each time, this "optimistic" approach is more appropriate (as in most cases it does not need to be called, even saving the allocation of PingPong). 

Of course, I would much prefer that the owner of this repo natively supports this feature. I think this feature is a very commonly used one. CC @ihciah @Lzzzzzt

**Comment by ethe** [comment]
`futures_lite::yield_now` will be working on monoio after release of #362
